File: James/230517_study/프로_152996_시소짝꿍.java
import java.util.*;

class Solution {
    static int[][] distances = {{1, 1}, {3, 2}, {4, 2}, {4, 3}};
    
    public int upperSearch(int target, int[] weights, int start, int end) {
        while(start < end) {
            int mid = (start + end) / 2;
            
            if(target < weights[mid]) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }
        
        return start;
    }
    
    public int lowerSearch(int target, int[] weights, int start, int end) {
        while(start < end) {
            int mid = (start + end) / 2;
            
            if(target <= weights[mid]) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }
        
        return start;
    }
    
    public long solution(int[] weights) {
        long answer = 0;
        
        Arrays.sort(weights);
        
        for(int[] distance : distances) {
            for(int i = 0; i < weights.length; i++) {
                int x = weights[i];
                
                if(x * distance[0] % distance[1] != 0)
                    continue;
                
                int y = (distance[0] * x) / distance[1];
                
                int upper = upperSearch(y, weights, i + 1, weights.length);
                int lower = lowerSearch(y, weights, i + 1, upper);
                
                answer += upper - lower;
            }
        }
        
        return answer;
    }
}

File: sammy/230517_study/프로_152996_시소 짝꿍.py
from collections import Counter

def solution(weights):
    answer = 0
    countList=Counter(weights)

    for key,value in countList.items():
        if  value>=2: # 해당 무게를 가진 사람이 두명 이상이면 짝꿍 만들기 가능.
            answer += value*(value-1)//2
        weights=set(weights) # 중복 없애면 빨라짐.
        if key*2/3 in weights: # 2:3 비율을 가진 값 찾기
            answer += countList[key*2/3] * countList[key]

        if key*2/4 in weights: # 2:4 비율을 가진 값 찾기
            answer += countList[key*2/4] * countList[key]
        
        if key*3/4 in weights: # 3:4 비율을 가진 값 찾기
            answer += countList[key*3/4] * countList[key]

    return answer

print(solution([100,180,360,100,270]),4)